File: csharp/components/Strata.Application/Core/AbstractServiceEventSource.cs
﻿//  ##########################################################################
//  # File Name: AbstractServiceEventSource.cs
//  # Copyright: 2018, Sapientia Systems, LLC.
//  ##########################################################################

using Strata.Foundation.Utility;
using System.Collections.Generic;

namespace Strata.Application.Core
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// </summary>
    ///  
    public abstract
    class AbstractServiceEventSource<S,E,O>:
        IServiceEventSource<S,E,O>
        where S: IServiceEventSource<S,E,O>
        where E: IServiceEvent<S>
        where O: IServiceEventObserver<E>
    {
        public virtual ISet<O> Observers { get; protected set; }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        ///  
        protected
        AbstractServiceEventSource()
        {
            Observers = new HashSet<O>();
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        ///  
        public virtual S 
        AttachFrom(S other)
        {
            return Attach(other.Observers);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        ///  
        public virtual S
        Attach(ISet<O> observers)
        {
            observers.ForEach((O observer) => Observers.Add(observer));
            return GetSelf();
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        ///  
        public virtual S
        Attach(O observer)
        {
            Observers.Add(observer);
            return GetSelf();
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        ///  
        public virtual S
        Detach(O observer)
        {
            Observers.Remove(observer);
            return GetSelf();
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        ///  
        public virtual bool
        Has(O observer)
        {
            return Observers.Contains(observer);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        ///  
        public virtual S
        Notify(E evnt)
        {
            Observers.ForEach((O observer) => observer.OnEvent(evnt));
            return GetSelf();
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        ///  
        protected abstract S
        GetSelf();
    }
}

//  ##########################################################################

File: java/Components/Strata.Application.Core/src/strata/application/core/service/AbstractServiceEventSource.java
//////////////////////////////////////////////////////////////////////////////
// AbstractServiceEventSource.java
//////////////////////////////////////////////////////////////////////////////

package strata.application.core.service;

import java.util.HashSet;
import java.util.Set;

/*****************************************************************************
 * Base implementation of the {@code IServiceEventSource<S,E,O>} interface.
 *
 * @param <S> service event source type
 * @param <E> service event type
 * @param <O> service event observer type
 */
public abstract
class AbstractServiceEventSource<
    S extends IServiceEventSource<S,E,O>,
    E extends IServiceEvent<S>,
    O extends IServiceEventObserver<E>>
    implements IServiceEventSource<S,E,O>
{
    private final Set<O> observers;

    /*************************************************************************
     * Creates  a new instance of {@code AbstractServiceEventSource<S,E,O>}.
     */
    protected
    AbstractServiceEventSource()
    {
        observers = new HashSet<O>();
    }

    /*************************************************************************
     * {@inheritDoc}
     */
    @Override
    public S
    attachFrom(S other)
    {
        observers.addAll(other.getObservers());
        return getSelf();
    }

    /*************************************************************************
     * {@inheritDoc}
     */
    @Override
    public S
    attach(Set<O> observers)
    {
        observers.addAll(observers);
        return getSelf();
    }

    /*************************************************************************
     * {@inheritDoc}
     */
    @Override
    public S
    attach(O observer)
    {
        observers.add(observer);
        return getSelf();
    }

    /*************************************************************************
     * {@inheritDoc}
     */
    @Override
    public S
    detach(O observer)
    {
        observers.remove(observer);
        return getSelf();
    }

    /*************************************************************************
     * {@inheritDoc}
     */
    @Override
    public boolean
    has(O observer)
    {
        return observers.contains(observer);
    }

    /*************************************************************************
     * {@inheritDoc}
     */
    @Override
    public S
    notify(E event)
    {
        for (O observer:observers)
            observer.onEvent(event);

        return getSelf();
    }

    /*************************************************************************
     * Returns a reference to this.
     *
     * @return reference to this for method chaining
     */
    protected abstract S
    getSelf();
}

//////////////////////////////////////////////////////////////////////////////
